### PR TITLE
Add batch_norm decomposition for Spyre

### LINKED ISSUE #149: Enable torch.nn.functional.batch_norm
### What feature would you like to be added?

Enable torch.nn.functional.batch_norm on Spyre

### Why is this feature needed?

Allows models using this operator to work on Spyre

### Any alternative solution that you may want to share?

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by yoheiueda**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Implements custom batch normalization decomposition for Spyre based on upstream PyTorch's implementation.

- Based on [upstream PyTorch v2.10.0](https://github.com/pytorch/pytorch/blob/v2.10.0/torch/_decomp/decompositions.py#L1932-L1969)
- Operates in float16 (vs upstream's float32) due to pending type conversion support
    - See #937
- Uses `transpose` instead of `unsqueeze` to avoid restickify requirement
- Uses `eps=1e-4` in tests instead of default `1e-5` for float16 precision

#### Which issue(s) this PR is related to:

Fixes  #149 

#### Additional note:

```
(dt-inductor) [yohei@yohei-spyre-dev torch-spyre]$ pytest tests
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.12.12, pytest-9.0.2, pluggy-1.6.0
rootdir: /home/yohei/dt-inductor/torch-spyre
configfile: pyproject.toml
plugins: hypothesis-6.151.9
collected 460 items

tests/_inductor/test_building_blocks.py ....                                                                                                                     [  0%]
tests/_inductor/test_inductor_decomp.py ...                                                                                                                      [  1%]
tests/_inductor/test_inductor_ops.py ...s....................................................................................................................... [ 28%]
................................................................................................................................................................ [ 63%]
................................                                                                                                                                 [ 70%]
tests/_inductor/test_logging.py ....                                                                                                                             [ 70%]
tests/tensor/test_tensor_layout.py ...............                                                                                                               [ 74%]
tests/test_fallbacks.py ........                                                                                                                                 [ 75%]
tests/test_modules.py .sssss.                                                                                                                                    [ 77%]
tests/test_ops.py .x.s..xs...s........................s...sss.ss......................................                                                           [ 95%]
tests/test_regex.py .                                                                                                                                            [ 95%]
tests/test_spyre.py .s..ss............                                                                                                                           [ 99%]
tests/test_spyre_lazy_silent.py .                                                                                                                                [100%]

======================================================== 440 passed, 18 skipped, 2 xfailed in 411.34s (0:06:51) ========================================================
```

### CODE REVIEW

**Review by michihirohorie — COMMENTED** [review]
@yoheiueda I’m wondering if we can add a few tests under tests/inductor/test_inductor_decomp.py, since batch_norm_decomp expands a single op into various sub-ops. If I understand correctly, test_inductor_decomp.py can be used to verify the decomposition contract, that is, to ensure an op is rewritten into the expected set of sub-ops.

For example, we could assert that the graph contains the expected ops (add + sqrt + reciprocal) in the invstd chain. Also, we could also assert that unsqueeze does not appear in the graph. 

Does this make sense?

**Comment by yoheiueda** [comment]
The upstream decomposition for batch_norm relies on the following ops, all of which are already supported or will be supported soon:

* unsqueeze
* flatten
* scalar value
* sqrt / rsqrt for float32
* conversions between float16 and float32 

https://github.com/pytorch/pytorch/blob/v2.10.0/torch/_decomp/decompositions.py#L1932-L1969

So, I don't think we need to introduce a custom decomposition for batch_norm.
